### ISSUE TITLE
UR 3.X support 

### ISSUE THREAD

**Issue by robotlinker** [andyzeng/visual-pushing-grasping#11]
Hi 
I am using a UR5 robot version 3.5.1 to test the package and get errors, looks like it is only tested with UR Software version 1.8? Any suppor for Version 3.X using ROS ur_modern_driver?
Thanks!

**Comment by WwYyFan** [comment]
@robotlinker
Hi,I also encountered this problem. Can you solve it now?

**Comment by nouyang** [comment]
...What errors are you encountering?

 I was able to get it working with Version 3.x, there was a little finickiness with the parse_tcp_state (assert message type 16), which I hacked around (by duplicating any `tcp_socket.recv` lines). This is probably doing something like flushing out data until it's indexed into the right place, but I'm moving on for now. A more robust TCP URx library can be found at https://github.com/SintefManufacturing/python-urx ; I haven't really dug into the differences.

**Comment by kwang4372** [comment]
Hi, I have a problem with 'Robot' class init.
i modify a line ('assert(robot message type == 16) to 20. ( i have ur3 3.5)
but, i encounter another problem.
At return of parse_tcp_state_data, I have a error 
"struct.error : unpack requires a string argument of length 8"
Please, give me a help 
![스크린샷, 2019-07-31 18-12-30](https://user-images.githubusercontent.com/53513962/62199673-e5149400-b3be-11e9-9cc0-17de47486ae8.png)

![스크린샷, 2019-07-31 18-13-18](https://user-images.githubusercontent.com/53513962/62199681-e80f8480-b3be-11e9-8932-0a1d660f7657.png)

**Comment by liuxiangdi** [comment]
> Hi, I have a problem with 'Robot' class init.
> i modify a line ('assert(robot message type == 16) to 20. ( i have ur3 3.5)
> but, i encounter another problem.
> At return of parse_tcp_state_data, I have a error
> "struct.error : unpack requires a string argument of length 8"
> Please, give me a help
> ![스크린샷, 2019-07-31 18-12-30](https://user-images.githubusercontent.com/53513962/62199673-e5149400-b3be-11e9-9cc0-17de47486ae8.png)
> 
> ![스크린샷, 2019-07-31 18-13-18](https://user-images.githubusercontent.com/53513962/62199681-e80f8480-b3be-11e9-8932-0a1d660f7657.png)

Hi, Have you solve this ploblem?, I am using UR5 version3.5.4, i meet the same problem, I tried use UR script a API and it works, But I still can not use this code

**Comment by Abdul-UOiT** [comment]
@liuxiangdi how are you trying to connect with the real ur5
do you use ROS `ur_modern_driver`?

**Comment by Abdul-UOiT** [comment]
can anyone here explain how to connect with real robot? 
should it be connected through `ros vrep bridge`

**Comment by liuxiangdi** [comment]
> @liuxiangdi how are you trying to connect with the real ur5
> do you use ROS `ur_modern_driver`?

no, finally I use urx-python package for UR control (the parser code in this repo don't suit newest UR version)
[(https://github.com/SintefManufacturing/python-urx)]
for realsense, you can use code in this repo(just change an API in realsense reading code)

**Comment by Lightwingy** [comment]
@liuxiangdi , can you run the code in the real robot? If you can, how do you use urx-python package for UR control?

**Comment by liuxiangdi** [comment]
> @liuxiangdi , can you run the code in the real robot? If you can, how do you use urx-python package for UR control?

you can use code in this repo(need rewrite the UR parser part, in my robot version 3.5, this code parser is old version)
or you can use URX-python, and just delete all control code in this repo then change to urx-python.
It is not very difficult because this paper just have 4 direction basic move, you can use urx-python for control test, it is very easy because it provide high level API.e.
I run a similary task in UR5, i use code in this repo first, it can run but i did not test the pushing and grasping task because i don't have robotQ for grasping.

**Comment by Lightwingy** [comment]
@liuxiangdi Thank you!

**Comment by gsbandy24** [comment]
> @liuxiangdi Thank you!



> > @liuxiangdi , can you run the code in the real robot? If you can, how do you use urx-python package for UR control?
> 
> you can use code in this repo(need rewrite the UR parser part, in my robot version 3.5, this code parser is old version)
> or you can use URX-python, and just delete all control code in this repo then change to urx-python.
> It is not very difficult because this paper just have 4 direction basic move, you can use urx-python for control test, it is very easy because it provide high level API.e.
> I run a similary task in UR5, i use code in this repo first, it can run but i did not test the pushing and grasping task because i don't have robotQ for grasping.

Hi, it seems you two have figured out the solution here. I am currently experiencing the same issues with the repo and need to make the changes either to the parser or adapt the repo to the urx-python. 

In what way were you able to figure it out @Lightwingy
